### ISSUE TITLE
Is there a difference in accuracy of approximately 1e-6 between the CPU and GPU in PyTorch inference results

### ISSUE THREAD

**Issue by lzcchl** [pytorch/pytorch#121403]
Provide a short description.

#########################################
  import os
import torch
import torchvision
from torchvision import transforms
import cv2
from PIL import Image
import numpy as np

useGPU = True#True#False

def read_class_file(class_file_path):
    class_info = []
    with open(class_file_path, 'r') as f:
        for line in f:
            class_info.append(line)
    return class_info

def infer_res50(data_dir, class_info):
    #net
    net = torchvision.models.resnet.resnet50(weights='ResNet50_Weights.DEFAULT')
    if useGPU == True:
        net.cuda()
    net.eval()

    #data
    infer_transform = transforms.Compose([
        transforms.Resize([256,256]),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize(mean=[0.4848, 0.4435, 0.4023], std=[0.2744, 0.2688, 0.2757])
    ])

    with torch.no_grad():
        for file_name in os.listdir(data_dir):
            if file_name.endswith('.jpg'):
                file_path = os.path.join(data_dir, file_name)
                img = Image.open(file_path)
                input_tensor = infer_transform(img)

                #infer
                input_batch = input_tensor.unsqueeze(0)
                if useGPU == True:
                    input_batch = input_batch.cuda()
                output = net(input_batch)
                score, index = torch.max(output, -1)
                output_np = output.cpu().numpy()
                print('score: {0}, class: {1}'.format(score.cpu().numpy()[0], class_info[index]))

    print('infer_res50 over')

if __name__ == '__main__':
    data_dir = r'./dataset'
    class_file_path = r'./dataset/class.txt'
    class_info = read_class_file(class_file_path)
    infer_res50(data_dir, class_info)
    print('all over')
#########################################

- OS: ubuntu2004
- PyTorch version:2.1.2+cu121
- Python version:3.8
- CUDA/cuDNN version:12.1 /

I printed the output of the first layer of the network and the final output of the network, and I found that there is a difference in accuracy after the 6th bit of the floating-point number. Is this situation normal? As shown in the figure below

Is there any configuration for torch that can be changed so that the accuracy of the CPU and GPU can be completely consistent?

![0001](https://github.com/pytorch/pytorch/assets/34515022/eb7ea176-8afa-458b-be55-7d61abc44a8c)
![0002](https://github.com/pytorch/pytorch/assets/34515022/384ce189-912d-4938-8a4a-f30dd12e4925)

**Comment by IMbackK** [comment]
The gpu and cpu paths (and hardware) are going to have slightly different float characteristics. You cant really avoid this.
A pretty nuclear and invasive way to get cross architecture determinism would be to use integer operations only.
